core/src/main/java/tc/oc/pgm/consumable/ConsumableModule.java

```diff
@@ -0,0 +1,58 @@
+package tc.oc.pgm.consumable;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.logging.Logger;
+import org.jdom2.Document;
+import org.jdom2.Element;
+import org.jetbrains.annotations.Nullable;
+import tc.oc.pgm.api.map.MapModule;
+import tc.oc.pgm.api.map.factory.MapFactory;
+import tc.oc.pgm.api.map.factory.MapModuleFactory;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.module.exception.ModuleLoadException;
+import tc.oc.pgm.kits.Kit;
+import tc.oc.pgm.kits.KitNode;
+import tc.oc.pgm.util.xml.InvalidXMLException;
+import tc.oc.pgm.util.xml.Node;
+import tc.oc.pgm.util.xml.XMLUtils;
+
+public class ConsumableModule implements MapModule<ConsumableMatchModule> {
+
+  private Set<ConsumableDefinition> consumableDefinitions = new HashSet<>();
+
+  @Override
+  public @Nullable ConsumableMatchModule createMatchModule(Match match) throws ModuleLoadException {
+    return new ConsumableMatchModule(match, consumableDefinitions);
+  }
+
+  public static class Factory implements MapModuleFactory<ConsumableModule> {
+
+    @Override
+    public @Nullable ConsumableModule parse(MapFactory factory, Logger logger, Document doc)
+        throws InvalidXMLException {
+      ConsumableModule consumableModule = new ConsumableModule();
+
+      for (Element consumableElement :
+          XMLUtils.flattenElements(doc.getRootElement(), "consumables", "consumable")) {
+        String id = XMLUtils.getRequiredAttribute(consumableElement, "id").getValue();
+        boolean override = XMLUtils.parseBoolean(consumableElement.getAttribute("override"), true);
+        Kit kit = factory.getKits().parseKitProperty(consumableElement, "kit", KitNode.EMPTY);
+        ConsumeAction consumeAction =
+            XMLUtils.parseEnum(
+                Node.fromAttr(consumableElement, "action"),
+                ConsumeAction.class,
+                "consume action",
+                ConsumeAction.EAT);
+
+        ConsumableDefinition consumableDefinition =
+            new ConsumableDefinition(id, kit, consumeAction, override);
+
+        factory.getFeatures().addFeature(consumableElement, consumableDefinition);
+        consumableModule.consumableDefinitions.add(consumableDefinition);
```

Comment:
This is terrible practice, do not mutate the module's state.

Create a `Set<ConsumableDefinition>` for parsing, parse all the definitions and add them to the set, and only at the very end, create the `new ConsumableModule(definitions)`, and if possible (ie: no consumables are defined) skip creating it at all, and just return a null.
Also consider making the consumable module constructor create an immutable copy of the set to avoid anything mutating the state, as that would lead to incorrect behavior (ie: re-cycling to the same maps multiple times ending in differing behavior)

Reply:
This, and some of the other code convention issues, were copied from the projectile implementation, so that should probably be updated at some point too.